tests/php/Extension/Publishable/PublishableSiteTreeTest.php

```diff
@@ -0,0 +1,330 @@
+        $this->assertCount(4, $toDelete);
+        // There should be 1 object being updated (our parent page)
+        $this->assertCount(1, $toUpdate);
+
+        // Check that the expected Pages are represented in our objectsToDelete() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedDeleteIds, $this->getIdsFromArray($toDelete));
+        // Check that the expected Pages are represented in our objectsToUpdate() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedUpdateIds, $this->getIdsFromArray($toUpdate));
+    }
+
+    public function testObjectsActionUnpublishRecursiveInclusion(): void
+    {
+        // Test that recursive parents and children
+        SiteTree::config()->set('regenerate_parents', PublishableSiteTree::RELATION_INCLUDE_RECURSIVE);
+        // Regardless of configuration, we expect children to be added in objectsToDelete() on unpublish action
+        SiteTree::config()->set('regenerate_children', PublishableSiteTree::RELATION_INCLUDE_RECURSIVE);
+
+        $page = $this->objFromFixture(SiteTree::class, 'page3');
+        $childPage = $this->objFromFixture(SiteTree::class, 'page4');
+        $grandchildPage = $this->objFromFixture(SiteTree::class, 'page5');
+        $virtualPage = $this->objFromFixture(VirtualPage::class, 'page1');
+        $parentPage = $this->objFromFixture(SiteTree::class, 'page2');
+        $grandparentPage = $this->objFromFixture(SiteTree::class, 'page1');
+
+        // We expect the following pages to be deleted
+        $expectedDeleteIds = [
+            $page->ID,
+            $childPage->ID,
+            $grandchildPage->ID,
+            $virtualPage->ID,
+        ];
+
+        // We expect the following pages to be updated
+        $expectedUpdateIds = [
+            $parentPage->ID,
+            $grandparentPage->ID,
+        ];
+
+        $context = [
+            'action' => 'unpublish',
+        ];
+
+        $toDelete = $page->objectsToDelete($context);
+        $toUpdate = $page->objectsToUpdate($context);
+
+        // There should be the same 4 pages being deleted
+        $this->assertCount(4, $toDelete);
+        // There should be 2 object being updated (our parent and grandparent pages)
+        $this->assertCount(2, $toUpdate);
+
+        // Check that the expected Pages are represented in our objectsToDelete() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedDeleteIds, $this->getIdsFromArray($toDelete));
+        // Check that the expected Pages are represented in our objectsToUpdate() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedUpdateIds, $this->getIdsFromArray($toUpdate));
+    }
+
+    public function testUrlsToCache()
+    {
+        // Page class is required because RedirectorPage extends Page
+        if (!class_exists(Page::class)) {
+            $this->markTestSkipped('This unit test requires the Page class');
+        }
+        $page = new Page(['Title' => 'MyPage']);
```

Comment:
```suggestion
        $page = Page::create();
        $page->Title = 'MyPage';
```
Please use injector and explicit properties as it makes the code maintenance easier. Applies to multiple instances in this change-set.

Reply:
We tend to avoid using injector in tests unless the test needs it for a few reasons:
1. If something breaks with injector, we don't want every single unit test to fail, because that makes it hard to see that it's actually just injector that's broken
2. It means everyone running the test will have the same results, even if they've got some temporary or project-specific injector config in their local environment
3. It's a lot clearer that this test doesn't have any magic going on that relies on injection - we just need a page instance.